tc/autotuner/cuda/autotuner.cc

```diff
@@ -0,0 +1,174 @@
+    const typename CudaBackend::MappingOptionsType& options,
+    const TuningConfiguration& configuration) {
+  TuningConfiguration conf = configuration;
+  conf.fromCudaMappingOptions(options);
+  return conf;
+}
+
+// This function is ran on a single pre-determined GPU, in a single thread
+// It takes the input/output DLTensor objects that reside on that GPU
+//
+// We pass the bestTimeSoFar as an option to avoid taking locks in this
+// function. This trades off a bit of conservativeness for code sanity.
+//
+// The function returns true if purning is possible and we can skip poorly
+// performing versions early.
+template <>
+bool skipExecutionOrWarmup<CudaBackend>(
+    typename CudaBackend::ExecutorType& executor,
+    const std::vector<const DLTensor*>& outputs,
+    const std::vector<const DLConstTensor*>& inputs,
+    size_t bestTimeSoFar) {
+  // 1. Prune based on the number of threads: if you don't hit at least k warps
+  // (default k = 8; 256 total threads, controlled by
+  // FLAGS_tuner_min_launch_total_threads) then it's likely the kernel is not
+  // performing great.
+  // This may be completely off but is a good first initial rule of thumb
+  // for stress-testing autotuning.
+  auto debugTuner = FLAGS_debug_tuner;
+  auto minThreads = FLAGS_tuner_min_launch_total_threads;
+  USING_MAPPING_SHORT_NAMES(BX, BY, BZ, TX, TY, TZ);
+  auto block = executor.block();
+  auto nThreads =
+      mappingSize(TX, block) * mappingSize(TY, block) * mappingSize(TZ, block);
+  auto grid = executor.grid();
+  auto nBlocks =
+      mappingSize(BX, grid) * mappingSize(BY, grid) * mappingSize(BZ, grid);
+  if (nBlocks * nThreads < minThreads) {
+    LOG_IF(INFO, debugTuner)
+        << "Skip configuration: too few threads " << grid << " / " << block;
+    return true;
+  } else {
+    LOG_IF(INFO, debugTuner)
+        << "Run configuration launch bounds blocks: " << grid
+        << " and threads: " << block << "\n";
+  }
+
+  // 2. Perform a first run which may have one of 2 behaviors:
+  //   2.a. return a very slow first execution time, we should stop
+  //     early. This is akin to pruning but in this case we have run once,
+  //   2.b. return a reasonable execution time, in which case we proceed with
+  //     warmup.
+  auto timings = executor.profile(inputs, outputs);
+  // 2.a.
+  constexpr size_t kCatastrophicPerfFactor = 100;
+  if (bestTimeSoFar < std::numeric_limits<size_t>::max() and
+      timings.kernelRuntime >= std::chrono::microseconds(
+                                   (kCatastrophicPerfFactor * bestTimeSoFar))) {
+    return true;
+  }
+  // 2.b. during autotuning we don't want to spend too much time executing,
+  // use a reduced number of iterations for warmup.
+  constexpr int kReducedWarmupIterations = 2;
+  for (size_t i = 1; i < kReducedWarmupIterations - 1; ++i) {
+    executor.profile(inputs, outputs);
+  }
+
+  // 3. After reasonable warmup, look at the performance and prune if
```

Comment:
is 2 iterations a reasonable warmup?

Reply:
this is for ranking not absolute perf, I haven't seen cases where 2 is too small but I haven't conducted a statistical study either so YMMV